checksum/luhn.go

```diff
@@ -0,0 +1,26 @@
+// lunh.go
+// description: Luhn algorithm
+// details: is a simple checksum formula used to validate a variety of identification numbers, such as credit card numbers, IMEI numbers, etc [Lunh](https://en.wikipedia.org/wiki/Luhn_algorithm)
+// author(s) [red_byte](https://github.com/i-redbyte)
+// see lunh_test.go
+
+// Package checksum describes algorithms for finding various checksums
+package checksum
+
+// LuhnAlgorithm This function calculates the checksum using the Luna algorithm
+func LuhnAlgorithm(s string) bool {
+	n := len(s)
+	number := 0
+	result := 0
+	for i := 0; i < n; i++ {
+		number = int(s[i]) - '0'
+		if i%2 == 0 {
+			number *= 2
+			if number > 9 {
+				number -= 9
+			}
+		}
+		result += number
+	}
+	return result%10 == 0
+}
```

Comment:
Early returns are better than nested `if` statements and try to keep the conversion to the user - keep the algorithm as simple as possible.

```suggestion
func LuhnAlgorithm(s []rune) bool {
	n := len(s)
	var number rune
	var result rune
	for i := 0; i < n; i++ {
		number = s[i] - '0'
		if i%2 != 0 {
			result += number
			continue
		}
		number *= 2
		if number > 9 {
			number -= 9
		}
		result += number
	}
	return result%10 == 0
}
```

Reply:
As for the early return, I agree, I corrected it, but I do not completely agree about the second part, IMHO we are operating in essence with numerical sequences and for this algorithm the array of runes, in my opinion, is redundant, the string is perfect here, in my opinion, but I'm ready to discuss this moment is with you.

Reply:
It's up to you here, very much a design preference. However, from my experience, numerical sequences should remain numerical in nature, that is rune is numerical type whereas a string is not a numerical type and rudimentary operations are much better operated using rune and byte. Consider this same algorithm in C++ or Java, you would not be using strings there but rather the char type because of the extra overhead strings have, same applies here imo but again, up to you.

Reply:
> 

changed the string to an array of runes